src/pudl/transform/eia861.py

```diff
@@ -193,79 +614,708 @@ def service_territory(raw_dfs, tfr_dfs):
+            This must include any dataframes from which we want to
+            compile BA-Util-State associations, which means this
+            function has to be called after all the basic transform
+            functions that depend on only a single raw table.
+
+    Returns:
+        dict: a dictionary of transformed dataframes. This function
+        both compiles the association table, and finishes the
+        normalization of the balancing authority table. It may be that
+        once the harvesting process incorporates the EIA 861, some or
+        all of this functionality should be pulled into the phase-2
+        transform functions.
+
+    """
+    # The dataframes from which to compile BA-Util-State associations
+    other_dfs = [
+        tfr_dfs["sales_eia861"],
+        tfr_dfs["demand_response_eia861"],
+        tfr_dfs["advanced_metering_infrastructure_eia861"],
+        tfr_dfs["dynamic_pricing_eia861"],
+    ]
+
+    logger.info("Building an EIA 861 BA-Util-State association table.")
+
+    old_date_ba_util = (
+        tfr_dfs["balancing_authority_eia861"]
+        .query("report_date<='2012-12-31'")
+        .loc[:, [
+            "report_date",
+            "balancing_authority_id_eia",
+            "utility_id_eia",
+        ]]
+        .dropna()
+        .drop_duplicates()
+    )
+
+    old_date_util_state = pd.DataFrame()
+    for df in other_dfs:
+        tmp_df = (
+            df.query("report_date<='2012-12-31'")
+            .loc[:, [
+                "report_date",
+                "utility_id_eia",
+                "state",
+            ]]
+            .dropna()
+            .drop_duplicates()
+        )
+        old_date_util_state = (
+            pd.concat([old_date_util_state, tmp_df])
+            .drop_duplicates()
+        )
+    old_date_ba_util_state = pd.merge(
+        old_date_ba_util, old_date_util_state, how="outer")
+
+    new_ba_code_id = (
+        tfr_dfs["balancing_authority_eia861"]
+        .query("report_date>='2013-01-01'")
+        .loc[:, [
+            "report_date",
+            "balancing_authority_code_eia",
+            "balancing_authority_id_eia",
+        ]]
+        .dropna()
+        .drop_duplicates()
+    )
+    new_date_ba_code_util_state = pd.DataFrame()
+    for df in other_dfs:
+        tmp_df = (
+            df.query("report_date>='2013-01-01'")
+            .loc[:, [
+                "report_date",
+                "balancing_authority_code_eia",
+                "utility_id_eia",
+                "state",
+            ]]
+            .dropna()
+            .drop_duplicates()
+        )
+        new_date_ba_code_util_state = (
+            pd.concat([new_date_ba_code_util_state, tmp_df])
+            .drop_duplicates()
+        )
+    new_date_ba_util_state = (
+        new_date_ba_code_util_state
+        .merge(new_ba_code_id, how="outer")
+        .drop("balancing_authority_code_eia", axis="columns")
+        .drop_duplicates()
+    )
```

Comment:
these new and old sections look very similar. I'd consider making a little generic local function.